scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordbufs/TreeOrderedBuf.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ Copyright 2014 Twitter, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.macros.impl.ordbufs
+
+import scala.reflect.macros.Context
+import scala.language.experimental.macros
+
+import com.twitter.scalding._
+import java.nio.ByteBuffer
+import com.twitter.scalding.typed.OrderedBufferable
+
+object TreeOrderedBuf {
+  def toOrderedBufferable[T](c: Context)(t: TreeOrderedBuf[c.type])(implicit T: t.ctx.WeakTypeTag[T]): t.ctx.Expr[OrderedBufferable[T]] = {
+    import t.ctx.universe._
+    t.ctx.Expr[OrderedBufferable[T]](q"""
+      new _root_.com.twitter.scalding.typed.OrderedBufferable[$T] {
+
+        private[this] def innerCompare(a: _root_.java.nio.ByteBuffer, b: _root_.java.nio.ByteBuffer): Int = {
+            val ${t.compareBinary._1} = a
+            val ${t.compareBinary._2} = b
+            ${t.compareBinary._3}
+            return 0
+          }
+
+        def compareBinary(a: _root_.java.nio.ByteBuffer, b: _root_.java.nio.ByteBuffer): _root_.com.twitter.scalding.typed.OrderedBufferable.Result = {
+            val r = innerCompare(a, b)
+             if (r < 0) {
+                _root_.com.twitter.scalding.typed.OrderedBufferable.Less
+              } else if (r > 0) {
+                _root_.com.twitter.scalding.typed.OrderedBufferable.Greater
+              } else {
+                _root_.com.twitter.scalding.typed.OrderedBufferable.Equal
+              }
+          }
+
+        def hash(passedInObjectToHash: $T): Int = {
+          val ${t.hash._1} = passedInObjectToHash
+          ${t.hash._2}
+        }
+
+        def get(from: _root_.java.nio.ByteBuffer): _root_.scala.util.Try[(_root_.java.nio.ByteBuffer, $T)] = {
+          val ${t.get._1} = from
+          _root_.scala.util.Try {
+            (${t.get._1}, ${t.get._2})
+          }
+        }
+        def put(into: _root_.java.nio.ByteBuffer, t: $T): _root_.java.nio.ByteBuffer =  {
+          val ${t.put._1} = into
+          val ${t.put._2} = t
+          ${t.put._3}
+          ${t.put._1}
+        }
+
+        private def innerMemCompare(x: $T, y: $T): Int = {
+          val ${t.compare._1} = x
+          val ${t.compare._2} = y
+          ${t.compare._3}
+          return 0
+        }
+
+         def compare(x: $T, y: $T): Int = {
+          val tmp = innerMemCompare(x, y)
```

Comment:
this seems wasteful, if actually all of the compares except like short and byte are doing this. Seems like we should special case those.